### PR TITLE
[Product CRUD] Variable, variation, notices, and stock handling

### PR DESCRIPTION
**Pull request by mikejolley**
This covers several things, sorry for size.

- Variable product class methods
- Variation class methods
- Syncing between variable and variation, including deferred handling so it doesn't get done all at once per variation save.
- Admin panel uses getters
- Fixed several notices
- Improved upsell/cross sells to use the child ids rather than entire WP_Query objects.

### CODE REVIEW

**Comment by claudiosanches** [comment]
Looks good to me.

File: includes/abstracts/abstract-wc-product.php [hunk 1]
```diff
@@ -1537,16 +1609,22 @@ protected function update_post_meta() {
 			}
 		}
 
-		if ( $this->is_on_sale() ) {
-			update_post_meta( $this->get_id(), '_price', $this->get_sale_price() );
-		} else {
-			update_post_meta( $this->get_id(), '_price', $this->get_regular_price() );
+		if ( in_array( 'regular_price', $updated_props ) || in_array( 'sale_price', $updated_props ) ) {
```
Comment (justinshreve) [inline includes/abstracts/abstract-wc-product.php]: Shouldn't we also check for date settings changing here too? sale_price could be previously set, but someone could change and move up the sale date to 'today'.

File: includes/admin/class-wc-admin-post-types.php [hunk 2]
```diff
@@ -352,11 +352,11 @@ public function render_product_columns( $column ) {
 				echo $the_product->get_sku() ? $the_product->get_sku() : '<span class="na">&ndash;</span>';
 				break;
 			case 'product_type' :
-				if ( 'grouped' == $the_product->product_type ) {
+				if ( 'grouped' == $the_product->get_type() ) {
```
Comment (justinshreve) [inline includes/admin/class-wc-admin-post-types.php]: We can do is_type checks for these.

File: includes/class-wc-product-variable.php [hunk 3]
```diff
@@ -207,75 +202,22 @@ public function get_variation_default_attribute( $attribute_name ) {
...
-			return ( array_map( 'sanitize_title', $match_attributes ) === $match_attributes ) ? 0 : $this->get_matching_variation( array_map( 'sanitize_title', $match_attributes ) );
+	public function get_downloadable( $context = 'view' ) {
+		return false;
+	}
 
-		} else {
-			return 0;
-		}
+	/**
+	 * Variable products themselves cammpt be virtual.
```
Comment (justinshreve) [inline includes/class-wc-product-variable.php]: Small typo on "cannot".

File: includes/class-wc-product-variable.php [hunk 4]
```diff
@@ -310,440 +252,110 @@ public function get_available_variation( $variation ) {
...
-	 * Set stock status. This is synced with children (variations), not set directly.
-	 */
-	public function set_stock_status( $status ) {
-		$this->data['stock_status'] = 'outofstock' === $status ? 'outofstock' : 'instock';
-	}
+		$this->sync_managed_variation_stock_status();
+		$this->apply_changes();
+		$this->update_product_type();
+		$this->update_product_version();
+		$this->update_term_counts();
```
Comment (justinshreve) [inline includes/class-wc-product-variable.php]: Extra indent.

**Comment by justinshreve** [comment]
Left a few small comments but this is looking good and can be merged I think. 👍

**Pull request merged**